### PR TITLE
BEL-199 Basic outline of models in MVC discussion

### PR DESCRIPTION
**Pull request by tpaulshippy**
[Link to Jira ticket](https://strongmind.atlassian.net/browse/BEL-199)

## Purpose 
First draft of a discussion of writing good models in MVC frameworks like Rails.

## Approach 
Brainstormed ideas with Mark and wrote up a draft.

## Testing
n/a

## Screenshots/Video
n/a

### CODE REVIEW

**Review by markng — CHANGES_REQUESTED** [review]
- BEL-199 review  
	- I give this a 8 out of 10  
	- I like that  
		- Includes a purpose  
		- That it includes several examples of things that belong in the model's scope  
		- That it includes an understanding of what is in a controller  
			- and what "request logic" is  
		- that it calls out the idea of a maturity path  
		- that it includes smells  
	- To make it a 10  
		- it would call out that views represent the backend validation, rather than mirroring it  
		- it would give some examples in the maturity path  
		- it would include some examples of what a set of models might look like  
		- it would call out controllers that don't map to one resource/model as a smell  
		- it would call out controllers that have more than the basic CRUD methods  
			- (with the reminder that of what a smell actually is)

**Review by markng — APPROVED** [review]

File: tech_sessions/models.md [hunk 1]
```diff
@@ -0,0 +1,274 @@
+# Writing Good Models in MVC Frameworks
```
Comment (marc-hamilton-at-work) [inline tech_sessions/models.md:1]: Should the title of this be the same as what the user story says?

_Writing Good Models in Rails_

File: tech_sessions/models.md [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+2. Rich models with controllers that only route requests to the models.
+3. Models with common patterns refactored out into concerns.
+
+## Smells
+
+### Moving from Maturity Level 1 to 2
+* Controllers that contain business logic.
+* Controllers that don't map to one resource/model.
+* Controllers that have more than the basic CRUD actions.
+* Controllers that interact with models with more than 2-3 lines of code.
```
Comment (marc-hamilton-at-work) [inline tech_sessions/models.md:53]: I think I understand the point of this comment, but is there any relevance to the specific numbers of lines of code?

File: tech_sessions/models.md [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+
+### Moving from Maturity Level 2 to 3
+* Model methods that are more than 8-10 lines long.
```
Comment (marc-hamilton-at-work) [inline tech_sessions/models.md:56]: Same comment as above... is there any significance to the actual number of lines of code?

File: tech_sessions/models.md [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+* More than 3 or 4 methods in a model that relate to the same action.
+
+## Examples
```
Comment (marc-hamilton-at-work) [inline tech_sessions/models.md:59]: I'm wondering if it makes sense to start of with some very basic examples which do not have external dependencies?  (i.e. basic CRUD stuff against an underlying database)

File: tech_sessions/models.md [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+	end
+
+	def update_linkedin
+			LinkedIn.login(li_credentials)
+			LinkedIn.update(@shareable_text)
+	end
+end
+```
+
+### Model Association Examples
```
Comment (marc-hamilton-at-work) [inline tech_sessions/models.md:215]: I like that we're including examples of associations 👍

File: tech_sessions/models.md (hunk 1 above)
Comment (tpaulshippy) [reply tech_sessions/models.md:1]: We could. Most of this content is applicable in django and ASP.NET MVC also, but I'm happy to change the title.

File: tech_sessions/models.md (hunk 2 above)
Comment (tpaulshippy) [reply tech_sessions/models.md:53]: No, just a general guideline.

File: tech_sessions/models.md (hunk 3 above)
Comment (tpaulshippy) [reply tech_sessions/models.md:56]: No, just a ballpark.

File: tech_sessions/models.md (hunk 4 above)
Comment (tpaulshippy) [reply tech_sessions/models.md:59]: We could, although ActiveRecord kind of does CRUD for you so I'm not sure the model itself would be much different than what we did under `Model Association Examples`

To me, the problems generally come when you are moving beyond basic CRUD. And the tutorials all have that anyway, so I'm not sure about including it here.

**Pull request merged**